### DISCUSSION TITLE
Support for Deno packages

### DISCUSSION THREAD

**Discussion by amitav13** [vercel/turborepo#7454] [Ideas]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

We currently have supabase edge functions (written in TS) sitting in a package that we deploy independently of the rest of our Node/React packages and apps. We also have a `common` package that houses types and other utils that we use across our apps. We want to pull these in to our supabase functions as well.

### Describe the solution you'd like

I'd ideally want to just import the `common` package like we do in the rest of our apps, a line in package.json - 

`"common": "workspace:*"`

### Describe alternatives you've considered

A workaround would be to publish our package on npm and then add it in as a regular `esm.sh` import from a URL - though we lose all the goodness from having it as an internal dependency.

**Comment by tknickman** [comment]
Hey @amitav13, thanks for the issue. This is something that isn't immediately on the roadmap. Deno is pretty different in that there is no package manager or lockfile - everything is url based. Even the common standard of storing and exporting dependencies from `deps.ts` and `dev_deps.ts` is just a convention, and not enforced. This would make integrating turborepo difficult as the dependency paradigm is quite different. 

I'll leave this issue open so we can continue to gauge interest from the community, but as of now it's not something we're planning on in the short term.

**Reply by oneezy** [reply]
hey @tknickman , does this change now with **deno `v2`** ? i know `deno.lock` file's exist now along with workspaces / monorepo support but not sure what _"monorepo support"_ actually means here.

**Comment by amitav13** [comment]
> Hey @amitav13, thanks for the issue. This is something that isn't immediately on the roadmap. Deno is pretty different in that there is no package manager or lockfile - everything is url based. Even the common standard of storing and exporting dependencies from `deps.ts` and `dev_deps.ts` is just a convention, and not enforced. This would make integrating turborepo difficult as the dependency paradigm is quite different.
> 
> I'll leave this issue open so we can continue to gauge interest from the community, but as of now it's not something we're planning on in the short term.

Yeah I get that, its a hard one to pull in. I don't believe Nx or any other monorepo tool is supporting this atm. Would still be great to have though, we're just copy/pasting with a loud comment to remember to update in 2 places 😅

**Comment by lishaduck** [comment]
I'd like to note that my use case for a Deno Turbo is simpler, but still unsupported: I'd like a [Single-package workspaces](https://turbo.build/repo/docs/guides/single-package-workspaces), which wouldn't require dealing with workspaces. However, the [Multi-language support](https://turbo.build/repo/docs/guides/multi-language) guide doesn't work, as Deno has heuristics that are changed in the presence of a `package.json`. 
If we could, i.e., define tasks in `turbo.json`, or, better yet, use `tasks` from `deno.json`, then this would be sufficient for many Deno use cases.

Additionally, the upcoming Deno 2 will stabilize native workspace functionality, which may help monorepo use cases: denoland/deno#22942, and denoland/deno#24163 _might_ make a `package.json` for scripts feasible.

**Reply by lishaduck** [reply]
Deno 2.1 is out now, [with features based on Google's Wireit](https://deno.com/blog/v2.1#turbocharged-deno-task) (task dependencies, recursive runs) and plans to take on Turborepo natively soon (with caching).

**Reply by iharuya** [reply]
Deno currently [implements it on its own](https://github.com/denoland/deno/pull/26467/files).
And yeah I hope we have [caching as discussed here](https://github.com/denoland/deno/issues/26462) in the future :)

**Reply by lishaduck** [reply]
> And yeah I hope we have [caching as discussed here](https://github.com/denoland/deno/issues/26462) in the future :)

For future reference, I think the canonical issue for caching is https://github.com/denoland/deno/issues/26993.

**Comment by oneezy** [comment]
deno v2 support monorepos now. does this mean it will work with deno? i havent attempted it .. just doing some research.

**Reply by lishaduck** [reply]
Unfortunately, no. If you want to use Deno's Node-compatibility, you can, but if you want to operate Deno-native, you can't.

**Reply by oneezy** [reply]
**EDIT:** just read up on everything a bit more in depth. sounds like we gotta wait for deno to implement

@lishaduck that sucks! do you know what the issue is from stopping this from happening? turborepo or deno? also, do you know if there is  way to create a plugin for deno that bridges the gap or is it just something we have to wait for?

**Reply by lishaduck** [reply]
As [I said above](https://github.com/vercel/turborepo/discussions/7454#discussioncomment-11502066), the Turborepo team recommends using the Node-compat layer, while the Deno team recommends waiting on native Deno support.

**Reply by oneezy** [reply]
@lishaduck you see this?
https://x.com/JLarky/status/1859694001705844986

**Reply by lishaduck** [reply]
I'm not on X, but yeah. It doesn't support caching, but otherwise you can just use native Deno support.

**Reply by oneezy** [reply]
@lishaduck is this the magic bullet? 
> _"feat(unstable): support caching npm dependencies only as they're needed (https://github.com/denoland/deno/pull/27300)"_

![image](https://github.com/user-attachments/assets/5ad5f8ee-4c12-4d54-a608-5264917e434e)

**Reply by lishaduck** [reply]
> @lishaduck is this the magic bullet?
> 
> > _"feat(unstable): support caching npm dependencies only as they're needed ([denoland/deno#27300](https://github.com/denoland/deno/pull/27300))"_

Nope that PR, while welcome, just enables granular `"nodeModulesDir": "auto"` for "npm:*" dependencies.

**Comment by benjick** [comment]
What's the alternative today? I want to migrate to deno. What alternative to turborepo suits deno?
